sdk/helper/kdf/params.go

```diff
@@ -0,0 +1,206 @@
+	Method DKMLengthMethod
+
+	// Whether to encode the result in little endian or big endian.
+	LittleEndian bool
+
+	// Bit width of the specified variable; must be large enough to hold
+	// the value to encode; must be divisible by 8 with a maximum value of
+	// 64. Note that this is longer than CounterVariable.Width as it is not
+	// strictly specified by NIST. With 2^32 - 1 PRF invocations (max allowed
+	// by NIST), each generating 1024 bits (and generating 1024-bit keys),
+	// this could theoretically require a 42-bit counter, hence 64 seems a
+	// sane upper bound here.
+	Width uint8
+}
+
+// Defines a fixed byte array value (such as the info or context fields from
+// NIST SP800-108). No additional manipulation are performed on these fields.
+type ByteArray []byte
+
+// Encodes an integer value (as a uint64) into the specified byte width,
+// using the desired endianness.
+//
+// Note that all parameters above, like the PKCS#11 v3.0 standard, use _bits_
+// to specify the width; here we take _bytes_ to make the function easier to
+// read.
+func encodeInteger(value uint64, bytes uint8, little bool) []byte {
+	// Always allocate 8 bytes == 64 bits, our maximum possible size; we'll
+	// truncate it later as desired.
+	var result []byte = make([]byte, 8)
+
+	if little {
+		binary.LittleEndian.PutUint64(result, value)
+
+		// In little endian encoding, the value is left aligned; this means,
+		// to truncate to the specified width, we should remove the values
+		// thereafter.
+		result = result[:bytes]
+	} else {
+		binary.BigEndian.PutUint64(result, value)
+
+		// In big endian encoding, the value is right aligned; this means,
+		// to truncate to the specified width, we need to remove the leading
+		// values. To do so, we skip len(result) - $width bytes at the front,
+		// to leave $width bytes at the end.
+		result = result[8-bytes : 8]
+	}
+
+	return result
+}
+
+// Encodes, using the CounterVariable template, a given integer value.
+func (cv CounterVariable) Encode(value uint64) []byte {
+	return encodeInteger(value, cv.Width/8, cv.LittleEndian)
+}
+
+// Validates this parameter.
+func (cv CounterVariable) Validate() error {
+	if cv.Width == 0 {
+		return fmt.Errorf("potentially uninitialized CounterVariable: expected Width to be greater than zero; was 0")
+	}
+
+	if cv.Width%8 != 0 {
+		return fmt.Errorf("invalid value for CounterVariable: expected Width to be a multiple of 8; got %d", cv.Width)
+	}
+
+	if cv.Width > 32 {
+		return fmt.Errorf("invalid value for CounterVariable: expected Width to not exceed 32; got %d", cv.Width)
+	}
+
+	return nil
+}
+
+// Validates this parameter
+func (cv ChainingVariable) Validate() error {
+	return nil
+}
+
+func (kl DKMLength) CalculateDKMLength(prfSize int, keys []int) uint64 {
+	// Assumption: Validate has already passed, meaning there is exactly one
+	// of two values for kl.Method.
+	var result uint64
+
+	for _, keyBits := range keys {
+		if kl.Method == SumOfKeys {
```

Comment:
nit: Perhaps use a switch statement here? I'm always worried about additions/new value types to kl.Method?